### PR TITLE
[Central group table] Add membership manager to central group table 

### PR DESCRIPTION
**Pull request by mrow4a**
This is first approach to handle central group table feature. It utilizes new class Membership manager, which inherits from Access class. More implementation details will go inside the code. 

NOTE: Mind that this PR target is `centr_groups` branch, from this PR https://github.com/owncloud/core/pull/29107

@patrickjahns

### CODE REVIEW

**Comment by mrow4a** [comment]
I will also start writing unit tests to test queries I am using, so dont focus on them to much at that point. 

If we decide on that approach, lets merge it with https://github.com/owncloud/core/pull/29107. This should not affect any other class and existing functionalities. It will when all TODOs will be implemented.

File: lib/private/Group/Group.php [hunk 1]
```diff
@@ -64,6 +70,12 @@ class Group implements IGroup {
 	private $userManager;
 
 	/**
+	 * @var \OC\Group\BackendGroup $backendGroup
+	 */
+	private $backendGroup;
+
+	/**
+	 * // TODO: use BackendGroup here
```
Comment (mrow4a) [inline lib/private/Group/Group.php]: All the TODOs in the code are places which should be implemented after merging this PR

File: lib/private/Group/GroupMapper.php [hunk 2]
```diff
@@ -34,25 +38,27 @@ public function __construct(IDBConnection $db) {
...
 	public function getGroup($gid) {
 		$qb = $this->db->getQueryBuilder();
 		$qb->select('*')
 			->from($this->getTableName())
 			->where($qb->expr()->eq('group_id', $qb->createNamedParameter($gid)));
 
-		return $this->findEntity($qb->getSQL(), $qb->getParameters());
+		try {
+			/** @var BackendGroup $backendGroup */
+			$backendGroup = $this->findEntity($qb->getSQL(), $qb->getParameters());
```
Comment (mrow4a) [inline lib/private/Group/GroupMapper.php]: Add some additional checks which are not in original version. It should return null and not throw exception

File: lib/private/Group/Manager.php [hunk 3]
```diff
@@ -143,10 +144,21 @@ protected function clearCaches() {
 
 	/**
 	 * @param string $gid
+	 * @return boolean
+	 */
+	public function isCached($gid) {
```
Comment (mrow4a) [inline lib/private/Group/Manager.php]: This method is required to build optimized queries

File: lib/private/Server.php [hunk 4]
```diff
@@ -230,12 +231,20 @@ public function __construct($webRoot, \OC\Config $config) {
...
+		});
 		$this->registerService('AccountMapper', function(Server $c) {
-			return new AccountMapper($c->getConfig(), $c->getDatabaseConnection(), new AccountTermMapper($c->getDatabaseConnection()));
+			return new AccountMapper($c->getConfig(), $c->getDatabaseConnection(), $this->getAccountTermMapper());
 		});
 		$this->registerService('GroupMapper', function(Server $c) {
 			return new GroupMapper($c->getDatabaseConnection());
 		});
+		$this->registerService('MembershipManager', function(Server $c) {
+			return new MembershipManager($c->getDatabaseConnection(), $c->getConfig(),
```
Comment (mrow4a) [inline lib/private/Server.php]: All changes in this file are obligatory so it actually all works.

File: lib/public/AppFramework/Db/Mapper.php [hunk 5]
```diff
@@ -176,6 +176,23 @@ public function update(Entity $entity){
...
+	 *
+	 * If row contains invalid attributes, exception BadFunctionCallException will
+	 * be raised
+	 *
+	 * @param array $row the row which should be converted to an entity
+	 * @return Entity the entity
+	 * @throws \BadFunctionCallException
+	 * @since 7.0.0
+	 */
+	public function mapRowToEntity($row) {
```
Comment (mrow4a) [inline lib/public/AppFramework/Db/Mapper.php]: Make it public so MembershipManager can also create the Account entities (otherwise it will only be possible inside AccountMapper)

Added @throws in phpdoc to make it clear that this function should only be used by classes which operate over oc_accounts table.

File: lib/private/MembershipManager.php [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+use OC\Group\BackendGroup;
+use OC\User\Account;
+use OCP\AppFramework\Db\Access;
+use OCP\AppFramework\Db\Entity;
+use OCP\IConfig;
+use OCP\IDBConnection;
+use Doctrine\DBAL\Exception\UniqueConstraintViolationException;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+
+class MembershipManager extends Access {
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: Please have a look on all the public interfaces of that class.

File: lib/private/MembershipManager.php [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+	 * @param IQueryBuilder $qb
+	 * @param int $userId
+	 * @param string $gid
+	 * @return IQueryBuilder
+	 */
+	private function applyPredicates($qb, $userId, $gid) {
+		// Adjust the query depending on availability of accountId and $backendGroupId
+		// to have optimized access
+		$alias = $this->getTableAlias();
+		if ($this->userManager->isCached($userId) && $this->groupManager->isCached($gid)) {
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: I use isCached mathods of Managers to build optimized queries - it means, fetch info from cache if available, otherwise perform JOIN.

**Comment by mrow4a** [comment]
@jvillafanez

File: lib/private/MembershipManager.php (hunk 7 above)
Comment (jvillafanez) [reply lib/private/MembershipManager.php]: From my point of view, this reveals a design flaw. Why something related to the DB access has to check a class from the business logic? Why it needs to check if it's cached when noone should be interested in that?.

I think we should recheck the approach. What about the group manager handles also the membership? Could it simplify this?

**Comment by jvillafanez** [comment]
I'll need to do a second review

File: lib/private/Group/Manager.php (hunk 3 above)
Comment (mrow4a) [reply lib/private/Group/Manager.php]: True, will write a UserCacheManager and GroupCacheManager Włoch will be used Both in Group/User Manager and Mapper

**Comment by mrow4a** [comment]
Change is ready to review, full unit test coverage (target branch https://github.com/owncloud/core/pull/29107). @DeepDiver1975 @jvillafanez @butonic @PVince81

File: lib/private/MembershipManager.php [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+		$qb->selectAlias($qb->createFunction('COUNT(DISTINCT a.id)'), 'count');
+
+		$qb = $this->searchSQL($qb, $groupId, $isInternalGroupId, $pattern);
+
+		$stmt = $qb->execute();
+
+		// First should already contain count
+		$data = $stmt->fetch();
+		$stmt->closeCursor();
+		return intval($data['count']);
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: Is any case this might throw an exception in this query?

File: lib/private/MembershipManager.php [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+				'account_id' => $qb->createNamedParameter($accountId),
+				'membership_type' => $qb->createNamedParameter($membershipType),
+			]);
+
+		try {
+			$qb->execute();
+			return true;
+		} catch (UniqueConstraintViolationException $e) {
+			// FIXME: is this correct?
+			\OC::$server->getLogger()->logException($e, ['app' => 'core']);
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: @DeepDiver1975 is this correct way of logging this exception?

Comment (jvillafanez) [reply lib/private/MembershipManager.php]: We should check a couple of things:
* Does this class need to log something? Taking into account this is the only log here, probably not.
* Does this class need to handle that exception? I have my doubts about it.

The other option would be to not catch the exception and let it bubble up in order to handle it in other place. Note that exposing a DB exception and handle it near the output response might not be good.

I'm not saying this is bad, but there might be better options. My point is that having to log something without injecting a logger is something to be avoided, as well as injecting a class (the logger) that doesn't seem to be a clear dependency (because it's only used here). Note that swallowing the exception without doing anything is a worse option.

File: lib/private/MembershipManager.php [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+	 * @param bool $isInternalGroupId
+	 * @param string $pattern
+	 * @param integer $limit
+	 * @param integer $offset
+	 *
+	 * @return IQueryBuilder
+	 */
+	private function searchSQL($qb, $groupId, $isInternalGroupId, $pattern, $limit = null, $offset = null) {
+		$alias = $this->getTableAlias();
+		$allowMedialSearches = $this->config->getSystemValue('accounts.enable_medial_search', true);
```
Comment (jvillafanez) [inline lib/private/MembershipManager.php]: Is it worthy to extract this logic (getting the flag + setting the parameter variables) to another function? Taking into account this isn't used in other place than here, probably not unless the complexity of this function is high (I don't have data, but it doesn't seem so).

File: lib/private/MembershipManager.php [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+		} else if (!is_null($accountId)) {
+			// User predicate account_id specified
+			$qb->where($qb->expr()->eq('account_id', $qb->createNamedParameter($accountId)));
+		} else {
+			return false;
+		}
+
+		$qb->andWhere($qb->expr()->in('membership_type',
+			$qb->createNamedParameter($membershipTypeArray, IQueryBuilder::PARAM_INT_ARRAY)));
+		$qb->execute();
```
Comment (jvillafanez) [inline lib/private/MembershipManager.php]: What if the query fails? I expect the `execute()` function either return something or throws an exception. It doesn't seem this case is taken into account.

File: lib/private/Group/GroupMapper.php [hunk 12]
```diff
@@ -63,4 +76,18 @@ public function search($fieldName, $pattern, $limit, $offset) {
...
+	 * @param array $row the row which should be converted to backend group
+	 * @return BackendGroup the entity
+	 * @since 10.0
+	 *
+	 * @throws \BadFunctionCallException
+	 */
+	public function mapRowToEntity($row) {
+		/** @var BackendGroup $backendGroup */
+		$backendGroup = parent::mapRowToEntity($row);
+		return $backendGroup;
```
[the remark this answers is not in the document]
Comment (mrow4a) [reply lib/private/Group/GroupMapper.php]: Hmm, my point here was to make in transparent that the returned object is BackendGroup, not pure Entity. The same holds for Account. But yes, I could use this in the function which calls mapRowToEntity

**Comment by jvillafanez** [comment]
I haven't checked the DB queries yet. @mrow4a as unofficial DB expert I trust the queries work properly 😉  I can have a look if needed.

Btw, I don't have data about the test coverage of the MembershipManager, but it seems quite important to verify the coverage is near 100% in that class because we'll likely rely on it heavily and the DB queries aren't easy enough to ensure they'll work on all the cases in a glance.

File: lib/private/MembershipManager.php (hunk 9 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: Hmm, makes sense. In fact, we dont handle also foreign constrain violations, so maybe this one we should not either.

File: lib/private/MembershipManager.php (hunk 11 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: Hmm, true. Thanks!

File: lib/private/MembershipManager.php (hunk 10 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: Hmm, I have better idea what to do with it. Good point here.

File: lib/private/Group/GroupMapper.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (jvillafanez) [reply lib/private/Group/GroupMapper.php]: PHPDoc should take care of that, in this case in the parent class since it's the one that will be executed. Note that function visibility shouldn't need to be changed (from protected to public)

File: tests/lib/MembershipManagerTest.php [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+use OCP\AppFramework\Db\DoesNotExistException;
+
+/**
+ * Class MembershipManagerTest
+ *
+ * @group DB
+ *
+ * @package Test
+ */
+class MembershipManagerTest extends TestCase {
```
Comment (mrow4a) [inline tests/lib/MembershipManagerTest.php]: I covered 100% of the MembershipManager code here @jvillafanez, ale checked corner-cases, but it would be good to have a look if I did not overlook something.

File: lib/private/MembershipManager.php [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+	 * @param string|int $groupId
+	 * @param bool $isInternalGroupId
+	 * @param string $pattern
+	 * @return int
+	 */
+	private function countSqlQuery($groupId, $isInternalGroupId, $pattern) {
+		$qb = $this->db->getQueryBuilder();
+
+		// We need to use distinct since otherwise we will get duplicated rows for each search term
+		$qb->selectAlias($qb->createFunction('COUNT(DISTINCT a.`id`)'), 'count');
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: @butonic I needed a.`id` to make it work on Oracle, should I put it also in another DISTINCT?

**Comment by mrow4a** [comment]
100% unit tests coverage and tests passing on all dbms. PR ready for review and merging

File: lib/private/MembershipManager.php [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+			$qb->where($qb->expr()->eq('a.user_id', $qb->createNamedParameter($userId)));
+
+			// Apply predicate on group_id in backend groups table
+			$qb->andWhere($qb->expr()->eq('g.group_id', $qb->createNamedParameter($groupId)));
+		}
+
+		// Place predicate on membership_type
+		$qb->andWhere($qb->expr()->eq($alias.'.membership_type', $qb->createNamedParameter($membershipType)));
+
+		// Limit to 1, to prevent fetching unnecesary rows
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: Spelling - unnecessary

**Comment by jvillafanez** [comment]
Changes look good to me. 👍

File: lib/private/MembershipManager.php [hunk 16]
```diff
@@ -0,0 +1,714 @@
...
+	 * @param bool $isInternalGroupId
+	 * @param string $pattern
+	 * @param integer $limit
+	 * @param integer $offset
+	 *
+	 * @return Account[]
+	 */
+	private function searchAccountsSqlQuery($groupId, $isInternalGroupId, $pattern, $limit = null, $offset = null) {
+		$qb = $this->db->getQueryBuilder();
+		$qb->selectAlias('DISTINCT a.id', 'id')
```
Comment (mrow4a) [inline lib/private/MembershipManager.php]: Should I use DISTINCT a.`id` for Oracle @butonic?

File: lib/private/MembershipManager.php [hunk 17]
```diff
@@ -0,0 +1,714 @@
...
+	const MEMBERSHIP_TYPE_GROUP_USER = 0;
+	const MEMBERSHIP_TYPE_GROUP_ADMIN = 1;
+
+	/** @var IConfig */
+	protected $config;
+
+	/** @var IDBConnection */
+	private $db;
+
+	/** @var \OC\Group\GroupMapper */
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: import all classes in the use statement above

File: lib/private/MembershipManager.php [hunk 18]
```diff
@@ -0,0 +1,714 @@
...
+								\OC\User\AccountMapper $accountMapper, \OC\User\AccountTermMapper $termMapper,
+								\OC\Group\GroupMapper $groupMapper) {
+		$this->db = $db;
+		$this->config = $config;
+		$this->accountMapper = $accountMapper;
+		$this->groupMapper = $groupMapper;
+		$this->termMapper = $termMapper;
+	}
+
+	private function getTableAlias() {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: needed? just use 'm'

File: lib/private/MembershipManager.php [hunk 19]
```diff
@@ -0,0 +1,714 @@
...
+		return 'm';
+	}
+
+	private function getTableName() {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: needed= just use 'memberships'

File: lib/private/MembershipManager.php [hunk 20]
```diff
@@ -0,0 +1,714 @@
...
+	 * @return BackendGroup[]
+	 */
+	public function getAdminBackendGroups($userId) {
+		return $this->getBackendGroupsSqlQuery($userId, false, self::MEMBERSHIP_TYPE_GROUP_ADMIN);
+	}
+
+	/**
+	 * Return user account entities for given group (identified with gid)
+	 *
+	 * @param string $gid
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: string? not int?

File: lib/private/MembershipManager.php [hunk 21]
```diff
@@ -0,0 +1,714 @@
...
+	 *
+	 * @param string|int|null $groupId
+	 * @param bool $isInternalGroupId
+	 * @param int $membershipType
+	 * @return Account[]
+	 */
+	private function getAccountsSqlQuery($groupId, $isInternalGroupId, $membershipType) {
+		$qb = $this->db->getQueryBuilder();
+		$alias = $this->getTableAlias();
+		$qb->select(['a.id','a.user_id', 'a.lower_user_id', 'a.display_name', 'a.email', 'a.last_login', 'a.backend', 'a.state', 'a.quota', 'a.home'])
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: missing white space

File: lib/private/MembershipManager.php [hunk 22]
```diff
@@ -0,0 +1,714 @@
...
+		$qb->andWhere($qb->expr()->eq($alias.'.membership_type', $qb->createNamedParameter(self::MEMBERSHIP_TYPE_GROUP_USER)));
+
+		return $qb;
+	}
+
+	/**
+	 * @param IQueryBuilder $qb
+	 * @return int
+	 */
+	private function getAffectedQuery($qb) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: IQueryBuilder $qb

File: lib/private/MembershipManager.php [hunk 23]
```diff
@@ -0,0 +1,714 @@
...
+		// If affected is equal or more then 1, it means operation was successful
+		$affected = $qb->execute();
+		return $affected > 0;
+	}
+
+	/**
+	 * @param IQueryBuilder $qb
+	 * @return bool
+	 */
+	private function getExistsQuery($qb) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: IQueryBuilder $qb

File: lib/private/MembershipManager.php [hunk 24]
```diff
@@ -0,0 +1,714 @@
...
+	 * @param IQueryBuilder $qb
+	 * @param int $id
+	 * @param bool $isInternalId
+	 * @return IQueryBuilder
+	 */
+	private function applyUserPredicates($qb, $id, $isInternalId) {
+		// Adjust the query depending on availability of accountId
+		// to have optimized access
+		$alias = $this->getTableAlias();
+		if ($isInternalId) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: I'd split this into 2 functions - you gain nothing from this

File: lib/private/MembershipManager.php [hunk 25]
```diff
@@ -0,0 +1,714 @@
...
+	 * @param IQueryBuilder $qb
+	 * @param string $id
+	 * @param bool $isInternalId
+	 * @return IQueryBuilder
+	 */
+	private function applyGroupPredicates($qb, $id, $isInternalId) {
+		// Adjust the query depending on availability of accountId
+		// to have optimized access
+		$alias = $this->getTableAlias();
+		if ($isInternalId) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: split into two function

File: lib/private/User/AccountMapper.php [hunk 26]
```diff
@@ -168,7 +168,7 @@ public function find($pattern, $limit = null, $offset = null) {
 		$qb->selectAlias('DISTINCT a.id', 'id')
 			->addSelect(['user_id', 'lower_user_id', 'display_name', 'email', 'last_login', 'backend', 'state', 'quota', 'home'])
 			->from($this->getTableName(), 'a')
-			->leftJoin('a', 'account_terms', 't', $qb->expr()->eq('a.id', 't.account_id'))
```
Comment (DeepDiver1975) [inline lib/private/User/AccountMapper.php]: keep this for readability - we will never change the table names

File: tests/lib/MembershipManagerTest.php [hunk 27]
```diff
@@ -0,0 +1,626 @@
...
+	}
+
+	private static function clearDB() {
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*memberships`');
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*backend_groups`');
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*accounts`');
+	}
+
+	public static function tearDownAfterClass () {
+		MembershipManagerTest::clearDB();
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: self:: instead of MembershipManagerTest::

File: tests/lib/MembershipManagerTest.php [hunk 28]
```diff
@@ -0,0 +1,626 @@
...
+		$this->assertCount(4, $this->manager->getAdminAccounts());
+		$this->assertCount(4, $this->manager->getGroupUserAccounts($group->getGroupId()));
+	}
+
+	/**
+	 * Test that deleting group should result in deleting all users, and violating that
+	 * should rise exception
+	 */
+	public function testRemoveFailed() {
+		if ($this->connection->getDatabasePlatform() instanceof \Doctrine\DBAL\Platforms\SqlitePlatform) {
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: import sqliteplatform

File: tests/lib/MembershipManagerTest.php [hunk 29]
```diff
@@ -0,0 +1,626 @@
...
+			$this->markTestSkipped("Foreign key are not fully supported on SQLite, skip test");
+			return;
+		}
+		$account = MembershipManagerTest::getAccount(1);
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: self::

File: tests/lib/MembershipManagerTest.php [hunk 30]
```diff
@@ -0,0 +1,626 @@
...
+
+		$accountMapper = \OC::$server->getAccountMapper();
+		$groupMapper = \OC::$server->getGroupMapper();
+
+		$thrown = false;
+		try {
+			$accountMapper->delete($account);
+		} catch (\Exception $e) {
+			$thrown = true;
+			$this->assertInstanceOf(\Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException::class, $e);
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: import

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: lib/private/User/AccountMapper.php (hunk 26 above)
Comment (mrow4a) [reply lib/private/User/AccountMapper.php]: I exactly changed that for readability also, so we don't use magic strings which are easy to mistake or confuse

File: lib/private/MembershipManager.php (hunk 23 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: Not sure what you mean here

File: lib/private/MembershipManager.php (hunk 19 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: I would prefer to keep it, because we might use this table also in other parts of the code later - e.g. fs setup optimization.

Comment (mrow4a) [reply lib/private/MembershipManager.php]: I mean, your point is that table names won't change right?

Comment (DeepDiver1975) [reply lib/private/MembershipManager.php]: tablenames dont change - yes and a a one line function like this is pointless

File: lib/private/MembershipManager.php (hunk 23 above)
Comment (tomneedham) [reply lib/private/MembershipManager.php]: Type hint the parameter

File: lib/private/MembershipManager.php (hunk 14 above)
Comment (butonic) [reply lib/private/MembershipManager.php]: When using `$qb->createFunction(` you basically pass raw SQL. That is why you need to do the column escaping. With `$qb->expr()->eq(` that is not necessary, because it will do the column quoting (`a.id` becomes <code>a.&#96;id&#96;</code>) for you.

Where exactly would you put another DISTINCT?

File: lib/private/MembershipManager.php (hunk 16 above)
Comment (butonic) [reply lib/private/MembershipManager.php]: Distinc with alias is a littlve ugly. Whil doctrine has `selectAlias` and `selectDistinct` neither does what you need. `selectAlias` will treat `DISTINCT a.id` as a single identifier, escaping it to <code>DISTINCT a.`id`<code>. While that looks correct it is only by chance. Either you have test coverage for this, so we detect when docrine decides to treat this differently, or you wrap the rew sql in a function to make the problem explicit:
```php
$qb->selectAlias($qb->createFunction('DISTINCT a.`id`'), 'id');
```
Then again, this also only works as long as doctrine does not parse the funtion string for actual brackets...


So ... I would leave it as it is, but cover it with a unit test that checks the DISTINCT is working.

**Comment by mrow4a** [comment]
@DeepDiver1975 can we merge it into central_groups branch? I addressed all the issues you pointed

**Comment by mrow4a** [comment]
@DeepDiver1975 @PVince81 I have tested it with https://github.com/owncloud/core/pull/29282 integration into group manager, group, subadmin, user and user manager. This PR is ready

**Comment by mrow4a** [comment]
Any input here?

File: lib/private/MembershipManager.php [hunk 31]
```diff
@@ -0,0 +1,733 @@
...
+class MembershipManager {
+
+	/**
+	 * types of memberships in the group
+	 */
+	const MEMBERSHIP_TYPE_GROUP_USER = 0;
+	const MEMBERSHIP_TYPE_GROUP_ADMIN = 1;
+
+	/** @var IConfig */
+	protected $config;
```
Comment (jvillafanez) [inline lib/private/MembershipManager.php]: If this class isn't expected to be extended, this should be private.

File: lib/private/MembershipManager.php [hunk 32]
```diff
@@ -0,0 +1,733 @@
...
+
+	/** @var IDBConnection */
+	private $db;
+
+
+	public function __construct(IDBConnection $db, IConfig $config) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: PHPDoc missing

File: lib/private/MembershipManager.php [hunk 33]
```diff
@@ -0,0 +1,733 @@
...
+
+		return $this->getAffectedQuery($qb);
+	}
+
+	/*
+	 * Return backend group entities for given user id $userId (internal or gid) of which
+	 * the user has specific membership type
+	 *
+	 * @param string|int $userId
+	 * @param bool $isInternalUserId
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: what is internal?

File: lib/private/MembershipManager.php [hunk 34]
```diff
@@ -0,0 +1,733 @@
...
+		}
+
+		$qb->andWhere($qb->expr()->in('membership_type',
+			$qb->createNamedParameter($membershipTypeArray, IQueryBuilder::PARAM_INT_ARRAY)));
+
+		return $this->getAffectedQuery($qb);
+	}
+
+	/*
+	 * Return backend group entities for given user id $userId (internal or gid) of which
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: gid?

File: lib/private/MembershipManager.php [hunk 35]
```diff
@@ -0,0 +1,733 @@
...
+	}
+
+	/**
+	 * Return backend group entities for given account (identified by user's uid)
+	 *
+	 * @param string $userId
+	 *
+	 * @return BackendGroup[]
+	 */
+	public function getUserBackendGroups($userId) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: do we really need this function? From my understanding everything should internally work on accountIds

File: lib/private/Server.php [hunk 36]
```diff
@@ -991,6 +999,20 @@ public function getUserManager() {
...
+	 * @return \OC\MembershipManager
+	 */
+	public function getMembershipManager() {
+		return $this->query('MembershipManager');
+	}
+
+	/**
+	 * @return \OC\User\AccountTermMapper
+	 */
+	public function getAccountTermMapper() {
```
Comment (DeepDiver1975) [inline lib/private/Server.php]: why? public getters in class Server shall only return OCP interfaces

**Comment by DeepDiver1975** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! This pull request is being rebased automatically. Please DO NOT push while rebase is in progress or your changes would be lost!

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by ownclouders** [comment]
Automated rebase was successful!

File: lib/private/Server.php (hunk 36 above)
Comment (mrow4a) [reply lib/private/Server.php]: Hmm, so what with AccountMapper?

File: lib/private/MembershipManager.php (hunk 35 above)
Comment (mrow4a) [reply lib/private/MembershipManager.php]: It is just optimization, in case account id is only available by fetching the use from user manager.

**Comment by mrow4a** [comment]
@DeepDiver1975 addressed your doubts and fixed

File: tests/lib/MembershipManagerTest.php [hunk 37]
```diff
@@ -0,0 +1,631 @@
...
+
+use OC\Group\BackendGroup;
+use OC\Group\GroupMapper;
+use OC\User\Account;
+use OC\User\AccountMapper;
+use OC\MembershipManager;
+use OC\User\AccountTermMapper;
+use OCP\IConfig;
+use OCP\IDBConnection;
+use Test\TestCase;
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: unused

File: tests/lib/MembershipManagerTest.php [hunk 38]
```diff
@@ -0,0 +1,631 @@
...
+use OCP\AppFramework\Db\DoesNotExistException;
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: unused

File: tests/lib/MembershipManagerTest.php [hunk 39]
```diff
@@ -0,0 +1,631 @@
...
+		$this->manager = new MembershipManager(
+			$this->connection,
+			$this->config
+		);
+	}
+
+	private static function clearDB() {
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*memberships`');
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*backend_groups`');
+		\OC::$server->getDatabaseConnection()->executeQuery('DELETE FROM `*PREFIX*accounts`');
```
Comment (DeepDiver1975) [inline tests/lib/MembershipManagerTest.php]: really kill all?

File: lib/private/Server.php [hunk 40]
```diff
@@ -89,6 +89,7 @@
 use OCP\AppFramework\QueryException;
 use OCP\AppFramework\Utility\ITimeFactory;
 use OC\Group\GroupMapper;
+use OC\MembershipManager;
```
Comment (DeepDiver1975) [inline lib/private/Server.php]: unused

File: lib/private/Server.php [hunk 41]
```diff
@@ -610,13 +611,13 @@ public function __construct($webRoot, \OC\Config $config) {
...
 
 			return new Checker(
-					new EnvironmentHelper(),
-					new FileAccessHelper(),
-					new AppLocator(),
-					$config,
-					$c->getMemCacheFactory(),
-					$appManager,
-					$c->getTempManager()
+				new EnvironmentHelper(),
```
Comment (DeepDiver1975) [inline lib/private/Server.php]: remove whitspace changes

File: lib/private/User/AccountMapper.php [hunk 42]
```diff
@@ -239,4 +239,4 @@ public function callForAllUsers($callback, $search, $onlySeen) {
 		$stmt->closeCursor();
 	}
 
-}
+}
```
Comment (DeepDiver1975) [inline lib/private/User/AccountMapper.php]: remove whitespace changes

File: lib/private/MembershipManager.php [hunk 43]
```diff
@@ -0,0 +1,744 @@
...
+	/**
+	 * Add a group account (identified by user's internal account id)
+	 * to group (identified by group's internal backend group id).
+	 *
+	 * @param int $accountId - internal id of an account
+	 * @param int $backendGroupId - internal id of backend group
+	 *
+	 * @return bool
+	 */
+	public function addGroupUser($accountId, $backendGroupId) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: addGroupMemberSqlQuery throws UniqueConstraintViolationException - see PHPDoc
this method should do also throw it

File: lib/private/MembershipManager.php [hunk 44]
```diff
@@ -0,0 +1,744 @@
...
+	/**
+	 * Add a group admin account (identified by user's internal account id)
+	 * to group (identified by group's internal backend group id).
+	 *
+	 * @param int $accountId - internal id of an account
+	 * @param int $backendGroupId - internal id of backend group
+	 *
+	 * @return bool
+	 */
+	public function addGroupAdmin($accountId, $backendGroupId) {
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: same as with addGroupUser

File: lib/private/MembershipManager.php [hunk 45]
```diff
@@ -0,0 +1,744 @@
...
+	}
+
+	/**
+	 * Return account entities for given group
+	 * (identified by backend group id if $isBackendGroupId is true, or gid if $isBackendGroupId is false)
+	 * of which the accounts have specific membership type. If group id is not specified, it will
+	 * return result for all groups.
+	 *
+	 * @param string|int|null $groupId
+	 * @param bool $isInternalGroupId
```
Comment (DeepDiver1975) [inline lib/private/MembershipManager.php]: isInternalGroupId unknown 
isBackendGroupId missing in PHPDoc

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by mrow4a** [comment]
@DeepDiver1975 adjusted to comments

**Review by DeepDiver1975 — APPROVED** [review]
